File: packages/footballai/src/footballai/_paths.py
"""Shared path resolution for the FootballAI monorepo.

All scripts default to the monorepo root for shared runtime directories such as
``data/``, ``models/``, and ``external/``. The root is discovered by walking up
from this installed package and looking for the workspace manifest.
"""

from __future__ import annotations

from pathlib import Path


def find_repo_root() -> Path:
    """Return the monorepo root directory.

    The search starts at this file, walks upward looking for ``pyproject.toml``
    containing the uv workspace marker, and falls back to the first directory
    containing a ``.git`` folder.
    """
    here = Path(__file__).resolve()
    for parent in [here, *here.parents]:
        marker = parent / "pyproject.toml"
        if marker.exists():
            try:
                content = marker.read_text(encoding="utf-8")
                if "[tool.uv.workspace]" in content:
                    return parent
            except OSError:
                pass
        if (parent / ".git").is_dir():
            return parent
    # Fallback: package source root -> packages/footballai -> packages -> repo root
    return here.parent.parent.parent.parent


REPO_ROOT = find_repo_root()

File: apps/web/src/lib/server/paths.ts
import { existsSync, readFileSync } from 'node:fs';
import { dirname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

/**
 * Resolve the monorepo root from any file inside apps/web.
 *
 * The helper walks upward until it finds pyproject.toml containing the uv
 * workspace marker, then falls back to the first directory containing a .git
 * folder.
 */
export function workspaceRoot(): string {
	let current = dirname(fileURLToPath(import.meta.url));

	while (current !== dirname(current)) {
		const pyproject = resolve(current, 'pyproject.toml');
		if (existsSync(pyproject)) {
			try {
				const contents = readFileSync(pyproject, 'utf8');
				if (contents.includes('[tool.uv.workspace]')) {
					return current;
				}
			} catch {
				// ignore and keep walking
			}
		}
		if (existsSync(resolve(current, '.git'))) {
			return current;
		}
		current = dirname(current);
	}

	// Final fallback: four levels up from apps/web/src/lib/server/paths.ts
	return resolve(dirname(fileURLToPath(import.meta.url)), '..', '..', '..', '..');
}
